include/shell/shell.h

```diff
@@ -129,6 +129,41 @@ struct shell_static_entry {
 		.u = {.entry = &UTIL_CAT(_shell_, syntax)}		   \
 	}
 
+/**
+ * @brief Macro for defining and adding a conditional root command (level 0)
+ * with required number of arguments.
+ *
+ * @see SHELL_CMD_ARG_REGISTER for details.
+ *
+ * Macro can be used to create a command which can be conditionally present.
+ * It is and alternative to \#ifdefs around command registration and command
+ * handler. If command is disabled handler and subcommands are removed from
+ * the application.
+ *
+ * @param[in] flag	Compile time flag. Command is present only if flag
+ *			exists and equals 1.
+ * @param[in] syntax	Command syntax (for example: history).
+ * @param[in] subcmd	Pointer to a subcommands array.
+ * @param[in] help	Pointer to a command help string.
+ * @param[in] handler	Pointer to a function handler.
+ * @param[in] mandatory	Number of mandatory arguments.
+ * @param[in] optional	Number of optional arguments.
+ */
+#define SHELL_COND_CMD_ARG_REGISTER(flag, syntax, subcmd, help, handler, \
+					mandatory, optional) \
+	COND_CODE_1(\
+		flag, \
+		(\
+		SHELL_CMD_ARG_REGISTER(syntax, subcmd, help, handler, \
+					mandatory, optional) \
+		), \
+		(\
+		static shell_cmd_handler dummy_##handler \
+			__attribute__((unused)) = handler;\
+		static const struct shell_cmd_entry *dummy_subcmd_##syntax \
+			__attribute__((unused)) = subcmd\
+		)\
+	)
```

Comment:
Does the unused really means compiled out at the end? It is getting a bit heavy on macros, though not at level of logging, so I hope we don't follow with more of these.

Reply:
yes, unused makes compiler not complaining but linker removes it if it's unused.
It's getting heavy inside but imo it makes user life nicer. Like in case of the logger, macro magic makes log api simple to register modules and use with huge configurability underneath.

On the other hand, look at subsys/bluetooth/shell/bt.c and its #ifdef nightmare :)